### PR TITLE
integrating costume components and validation into the close loop

### PR DESCRIPTION
**Pull request by MoAljam**
I adjusted the workflow to fit to our adaptable experimentalist and added polynomial theorist alongside the linear for more comparisons.
we can get the validation results and progression given more data.

### CODE REVIEW

File: autora_closed_loop_demo/researcher_hub/autora_workflow.py [hunk 1]
```diff
@@ -21,39 +26,45 @@
...
-# - experimentalist
-# See more about components here: https://autoresearch.github.io/autora/
+    training_seq = trial_sequences(
+        coherence_ratios=[90, 10],
+        motion_directions=[45],
+        num_repetitions=1,
+        sequence_type="training",
+    )
 
+    print("len training sequence: ", len(training_seq[0]))
```
Comment (dilyuchik) [inline autora_closed_loop_demo/researcher_hub/autora_workflow.py:86]: maybe you want to comment out lines 59-63

File: autora_closed_loop_demo/researcher_hub/autora_workflow.py [hunk 2]
```diff
@@ -21,39 +26,45 @@
...
+
+    raw_data = pd.read_csv("myexperiment.csv")
+    return raw_data.to_dict(orient="records")
 
 
-state = StandardState(
-    variables=variables,
-)
+def run_experiment_once():
+    experiment_seq = trial_sequences(
```
Comment (dilyuchik) [inline autora_closed_loop_demo/researcher_hub/autora_workflow.py:72]: this and the line below should maybe be the coherence ratios and motion directions used in the actual experiment we're referencing, so:
coherence_ratios=[0, 5, 10, 20, 50]
motion_direction=[70, 160, 250, 340]

File: autora_closed_loop_demo/researcher_hub/autora_workflow.py [hunk 3]
```diff
@@ -81,140 +95,395 @@ def theorist_on_state(experiment_data, variables):
...
-    conditions_to_send['experiment_code'] = res
-    # upload and run the experiment, sending is to firestore database as conditions
-    data_raw = experiment_runner(conditions_to_send) # returns observations for each condition as jsPsych data
+
+    # global training_seq
+    # experiment_timeline = trial_sequences(coherence_ratios_list, motion_directions_list, all_items_in_one_trial=True)[0]
+    # js_code = stimulus_sequence(experiment_timeline, training_timeline=training_seq, to_html=False)
+
+    experiment_seq = trial_sequences(
+        coherence_ratios=[100, 50, 0],
```
Comment (dilyuchik) [inline autora_closed_loop_demo/researcher_hub/autora_workflow.py]: if we end up adjusting it based on the comment above, then this needs to be adjusted accordingly here too

File: autora_closed_loop_demo/researcher_hub/autora_workflow.py [hunk 4]
```diff
@@ -81,140 +95,395 @@ def theorist_on_state(experiment_data, variables):
...
+    )
+
+    training_seq = trial_sequences(
+        coherence_ratios=[90, 10],
+        motion_directions=[45],
+        num_repetitions=1,
+        sequence_type="training",
+    )
+
+    print("len training sequence: ", len(training_seq[0]))
```
Comment (dilyuchik) [inline autora_closed_loop_demo/researcher_hub/autora_workflow.py]: might want to comment out the print statements here as well

File: autora_closed_loop_demo/researcher_hub/autora_workflow.py [hunk 5]
```diff
@@ -81,140 +95,395 @@ def theorist_on_state(experiment_data, variables):
...
+
+    # *** Set up variables *** #
+    # independent variable is coherence in percent (0 - 100)
+    # dependent variable is rt in ms (0 - 10000)
+    variables = VariableCollection(
+        independent_variables=[
+            Variable(name="coherence_ratio", allowed_values=np.linspace(0, 100, 100), value_range=(0, 100)),
+            Variable(name="motion_direction", allowed_values=np.linspace(0, 360, 360), value_range=(0, 360)),
+        ],
+        dependent_variables=[Variable(name="d_prime", value_range=(0, 10000))],
```
Comment (dilyuchik) [inline autora_closed_loop_demo/researcher_hub/autora_workflow.py]: d_prime can actually also take negative values

**Review by dilyuchik — COMMENTED** [review]
I added a couple of small comments (in general, some print and display statements don't seem necessary anymore). I also don't think we need the play_ground folder on the main branch. But other than that, it looks good!

File: autora_closed_loop_demo/researcher_hub/autora_workflow.py (hunk 2 above)
Comment (MoAljam) [reply autora_closed_loop_demo/researcher_hub/autora_workflow.py:72]: good point

**Comment by MoAljam** [comment]
I resolved the few issues and structured it for running primarily online.

**Review by dilyuchik — APPROVED** [review]
Nice, thank you! The region setup at the beginning is a great addition

**Pull request merged**